File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/test_dashboard_coverage.py
#!/usr/bin/env python3

import json
import sys
import tempfile
import unittest
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE / "lib"))
import dashboard_coverage


SOURCE = """
<workbook>
  <dashboards>
    <dashboard name="Overview"><zones><zone id="1"/></zones></dashboard>
    <dashboard name="Operations"><zones><zone id="2"/></zones></dashboard>
    <dashboard name="Parameter Host"><zones><zone id="3"/></zones></dashboard>
    <dashboard name="Actually Empty"><zones/></dashboard>
  </dashboards>
  <windows>
    <window class="dashboard" name="Overview"/>
    <window class="dashboard" name="Operations"/>
    <window class="worksheet" hidden="true" name="Parameter Host"/>
  </windows>
</workbook>
"""


def spec(names):
    return {
        "name": "Fixture",
        "document": {
            "schemaVersion": 1,
            "kind": "workbook",
            "pages": [
                {"id": f"p{index}", "name": name}
                for index, name in enumerate(names)
            ],
            "elements": [],
            "layout": "",
        },
    }


class DashboardCoverageTest(unittest.TestCase):
    def setUp(self):
        self.temporary = tempfile.TemporaryDirectory()
        self.root = Path(self.temporary.name)
        self.twb = self.root / "source.twb"
        self.workbook = self.root / "spec.json"
        self.twb.write_text(SOURCE, encoding="utf-8")

    def tearDown(self):
        self.temporary.cleanup()

    def write_spec(self, names):
        self.workbook.write_text(json.dumps(spec(names)), encoding="utf-8")

    def test_full_scope_fails_on_missing_visible_dashboard(self):
        self.write_spec(["Overview"])
        result = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "full", "provenance": "full-workbook"},
        )
        self.assertEqual("fail", result["status"])
        self.assertEqual(["Operations"], result["missing_dashboards"])
        self.assertEqual(["Overview", "Operations"], result["visible_source_dashboards"])

    def test_full_scope_passes_when_complete(self):
        self.write_spec(["Overview", "Operations"])
        result = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "full", "provenance": "full-workbook"},
        )
        self.assertEqual("pass", result["status"])

    def test_only_stated_selected_scope_can_exclude_dashboard(self):
        self.write_spec(["Overview"])
        result = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "selected", "provenance": "stated", "dashboards": ["Overview"]},
        )
        self.assertEqual("pass", result["status"])
        self.assertEqual(["Operations"], result["scope_excluded_dashboards"])
        inferred = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "selected", "provenance": "inferred", "dashboards": ["Overview"]},
        )
        self.assertEqual("fail", inferred["status"])

    def test_story_points_are_required_pages_in_full_scope(self):
        story_plan = self.root / "story-plan.json"
        story_plan.write_text(
            json.dumps(
                [
                    {
                        "story": "Executive Story",
                        "points": [
                            {
                                "id": "1",
                                "caption": "Where we landed",
                                "captured_sheet": "Overview",
                                "sheet_kind": "dashboard",
                            }
                        ],
                    }
                ]
            ),
            encoding="utf-8",
        )
        self.write_spec(["Overview", "Operations"])
        missing = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "full", "provenance": "full-workbook"},
            story_plan,
        )
        self.assertEqual(["Where we landed"], missing["missing_story_points"])
        self.write_spec(["Overview", "Operations", "Where we landed"])
        complete = dashboard_coverage.evaluate(
            self.twb,
            self.workbook,
            {"mode": "full", "provenance": "full-workbook"},
            story_plan,
        )
        self.assertEqual("pass", complete["status"])


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/test-dashboard-coverage.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'tmpdir'
require_relative 'lib/dashboard_coverage'

fails = []
check = lambda do |condition, message|
  fails << message unless condition
  puts "  #{condition ? 'PASS' : 'FAIL'}  #{message}"
end

twb = <<~XML
  <workbook>
    <dashboards>
      <dashboard name="Overview"><zones><zone id="1"/></zones></dashboard>
      <dashboard name="Operations"><zones><zone id="2"/></zones></dashboard>
      <dashboard name="Parameter Host"><zones><zone id="3"/></zones></dashboard>
      <dashboard name="Actually Empty"><zones/></dashboard>
    </dashboards>
    <windows>
      <window class="dashboard" name="Overview"/>
      <window class="dashboard" name="Operations"/>
      <window class="worksheet" hidden="true" name="Parameter Host"/>
    </windows>
  </workbook>
XML

spec_for = lambda do |names|
  {
    'name' => 'Fixture',
    'document' => {
      'schemaVersion' => 1,
      'kind' => 'workbook',
      'pages' => names.map.with_index { |name, i| { 'id' => "p#{i}", 'name' => name } },
      'elements' => [],
      'layout' => ''
    }
  }
end

Dir.mktmpdir('dashboard-coverage') do |dir|
  twb_path = File.join(dir, 'source.twb')
  spec_path = File.join(dir, 'spec.json')
  File.write(twb_path, twb)

  File.write(spec_path, JSON.generate(spec_for.call(['Overview'])))
  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'full', 'provenance' => 'full-workbook' }
  )
  check.call(result['status'] == 'fail' && result['missing_dashboards'] == ['Operations'],
             'full-workbook mode fails when one visible Tableau dashboard is missing')
  check.call(result['visible_source_dashboards'] == %w[Overview Operations],
             'hidden parameter hosts and truly empty dashboards are excluded')

  File.write(spec_path, JSON.generate(spec_for.call(%w[Overview Operations])))
  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'full', 'provenance' => 'full-workbook' }
  )
  check.call(result['status'] == 'pass', 'full-workbook mode passes when every visible dashboard is built')

  File.write(spec_path, JSON.generate(spec_for.call(['Overview'])))
  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'selected', 'provenance' => 'stated', 'dashboards' => ['Overview'] }
  )
  check.call(result['status'] == 'pass' && result['scope_excluded_dashboards'] == ['Operations'],
             'stated selected scope may deliberately exclude another visible dashboard')

  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'selected', 'provenance' => 'inferred', 'dashboards' => ['Overview'] }
  )
  check.call(result['status'] == 'fail' &&
             result['blockers'].any? { |row| row['kind'] == 'unstated-scope' },
             'inferred selected scope cannot excuse a missing dashboard')

  story_plan = File.join(dir, 'story-plan.json')
  File.write(story_plan, JSON.generate([
    { 'story' => 'Executive Story',
      'points' => [{ 'id' => '1', 'caption' => 'Where we landed',
                     'captured_sheet' => 'Overview', 'sheet_kind' => 'dashboard' }] }
  ]))
  File.write(spec_path, JSON.generate(spec_for.call(%w[Overview Operations])))
  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'full', 'provenance' => 'full-workbook' },
    story_plan_path: story_plan
  )
  check.call(result['missing_story_points'] == ['Where we landed'],
             'full scope blocks when a Tableau story point has no Sigma page')
  File.write(spec_path, JSON.generate(spec_for.call(['Overview', 'Operations', 'Where we landed'])))
  result = DashboardCoverage.evaluate(
    twb_path: twb_path, spec_path: spec_path,
    scope: { 'mode' => 'full', 'provenance' => 'full-workbook' },
    story_plan_path: story_plan
  )
  check.call(result['status'] == 'pass', 'story point coverage passes after its page is built')
end

if fails.empty?
  puts 'ALL PASS'
  exit 0
end
warn "#{fails.length} FAILURE(S):"
fails.each { |failure| warn "  - #{failure}" }
exit 1
